Gru doesn't build with Lua 5.4 and above

Lua 5.4 introduced an API incompatibility (http://www.lua.org/manual/5.4/manual.html#8.3) that renders gru unable to build. lua_newuserdata() was replaced with lua_newuserdatauv().

The new function is defined as:
`void *lua_newuserdatauv (lua_State *L, size_t size, int nuvalue);`

It adds a third argument, `int nuvalue`, to the end of the function. Apparently setting this to 0 provides best performance, but sometimes it is necessary to set it to 1 for compatibility. Searching for this issue yields a few GitHub repos that have already made this change. See: https://github.com/ThePhD/sol2/issues/1052 and https://github.com/cloudwu/lua-bgfx/commit/40c404cdf5f0b86cef2f9bff7d0ed4dc98ced924

I imagine most major distros are still on Lua 5.3 or below, but some of the more leading-edge distros like Fedora are already on Lua 5.4, and all distros will eventually follow suit. Maybe it would be possible to allow compatibility with both Lua verisons using an ifdef on Lua 5.4 to define a macro (https://github.com/ThePhD/sol2/issues/1047#issuecomment-709078006) such as this one `#define lua_newuserdata(L, sz) lua_newuserdatauv((L), (sz), 0)`?

Thanks so much! I'd submit a PR myself but Lua is not my strong suit.

I tried replacing all 5 instances of `lua_newuserdata(L, sz)` with `lua_newuserdatauv((L), (sz), 1)` manually. Unfortunately that didn't actually change the error message I was receiving, which I realize I neglected to include in the original post.

> cd src/libgru && make all
> make[1]: Entering directory '/home/rylie/Downloads/git/n64/master/wii/src/libgru'
> gcc -I.. -I./.. -I./../../include  -std=gnu11 -Wall -g -O2  src/lib/blob.c src/lib/gru.c src/lib/gsc.c src/lib/n64rom.c src/lib/os.c src/lib/ups.c src/lib/util.c src/lib/z64fs.c src/exe/lblob.c src/exe/lgru.c src/exe/lgsc.c src/exe/ln64rom.c src/exe/los.c src/exe/lups.c src/exe/lz64fs.c src/exe/main.c ../../include/vector/vector.c -llua-5.3 -o gru
> /usr/bin/ld: /tmp/ccOlTpDv.o: in function `lgru_blob_create':
> /home/rylie/Downloads/git/n64/master/wii/src/libgru/src/exe/lblob.c:348: undefined reference to `lua_newuserdatauv'
> /usr/bin/ld: /tmp/ccwjUCky.o: in function `lgru_gsc_create':
> /home/rylie/Downloads/git/n64/master/wii/src/libgru/src/exe/lgsc.c:138: undefined reference to `lua_newuserdatauv'
> /usr/bin/ld: /tmp/ccGBRvox.o: in function `lgru_n64rom_create':
> /home/rylie/Downloads/git/n64/master/wii/src/libgru/src/exe/ln64rom.c:335: undefined reference to `lua_newuserdatauv'
> /usr/bin/ld: /tmp/cctHsFdy.o: in function `lgru_ups_create':
> /home/rylie/Downloads/git/n64/master/wii/src/libgru/src/exe/lups.c:122: undefined reference to `lua_newuserdatauv'
> /usr/bin/ld: /tmp/cccxDmrx.o: in function `lgru_z64fs_create':
> /home/rylie/Downloads/git/n64/master/wii/src/libgru/src/exe/lz64fs.c:788: undefined reference to `lua_newuserdatauv'
> collect2: error: ld returned 1 exit status
> make[1]: *** [makefile:27: gru] Error 1
> make[1]: Leaving directory '/home/rylie/Downloads/git/n64/master/wii/src/libgru'
> make: *** [makefile:48: all-libgru] Error 2

It seems like the suggested macro is already built-in to Lua 5.4 for backwards compatibility, but the "undefined reference to `lua_newuserdatauv` makes it sound like there's a missing declaration for that function in a header file? I've looked around in the gru header files, and can't seem to find an obvious place it needs to be declared.

--- a/configure
+++ b/configure
@@ -3914,7 +3914,7 @@ return luaopen_utf8 ();
   return 0;
 }
 _ACEOF
-for ac_lib in '' lua-5.3 lua5.3 lua; do
+for ac_lib in '' lua-5.4 lua5.4 lua-5.3 lua5.3 lua; do
   if test -z "$ac_lib"; then
     ac_res="none required"
   else
@@ -4243,7 +4243,7 @@ fi
 
 
 # Checks for header files.
-for ac_header in lua5.3/lua.h lua53/lua.h lua.h
+for ac_header in lua5.4/lua.h lua54/lua.h lua5.3/lua.h lua53/lua.h lua.h
 do :
   as_ac_Header=`$as_echo "ac_cv_header_$ac_header" | $as_tr_sh`
 ac_fn_c_check_header_mongrel "$LINENO" "$ac_header" "$as_ac_Header" "$ac_includes_default"

--- a/configure.ac
+++ b/configure.ac
@@ -76,7 +76,7 @@ AC_SUBST([LIBUTIL])
 AC_SUBST([LIBWS2_32])
 AC_SEARCH_LIBS([pow], [m],
                [AS_IF([test -n "$ac_lib"], [LIBM=-l$ac_lib])])
-AC_SEARCH_LIBS([luaopen_utf8], [lua-5.3 lua5.3 lua],
+AC_SEARCH_LIBS([luaopen_utf8], [lua-5.4 lua5.4 lua-5.3 lua5.3 lua],
                [AS_IF([test -n "$ac_lib"], [LIBLUA=-l$ac_lib])])
 AC_SEARCH_LIBS([libusb_init], [usb-1.0 usb],
                [AS_IF([test -n "$ac_lib"], [LIBUSB=-l$ac_lib])],
@@ -92,7 +92,7 @@ AC_SEARCH_LIBS([gai_strerrorA], [Ws2_32],
                [AS_IF([test -n "$ac_lib"], [LIBWS2_32=-l$ac_lib])])
 
 # Checks for header files.
-AC_CHECK_HEADERS([lua5.3/lua.h lua53/lua.h lua.h])
+AC_CHECK_HEADERS([lua5.4/lua.h lua54/lua.h lua5.3/lua.h lua53/lua.h lua.h])
 AC_CHECK_HEADERS([libusb-1.0/libusb.h libusb.h])
 AC_CHECK_HEADERS([pty.h util.h libutil.h])
 

